File: classes/Algoritmi/path/2020-03-26/piastrelle2.cpp
#include<iostream>
#include <cstdio>
#include<cassert>
using namespace std;
const int MAXN = 50; int N;

int p[2][MAXN];

int H(int n);
int G(int n);

int F(int n) {
  // ritorna il numero di piastrellature di un bagno 2xn ben squadrato.
  assert(n >= 0);
  if(n==0)
    return 1;
  if(n==1)
    return 2;
  return F(n-1) + G(n-1) + H(n-2);
}

int G(int n) {
  assert(n >= 0);
  if(n==0)
    return 1;
  return F(n) + G(n-1);
}

int H(int n) {
  assert(n >= 0);
  return F(n) + G(n);
}

void printP() {
  cout << "ciao N = " << N << endl; 
  for(int i=0; i<N; i++) {
//    assert( p[0][i] >= 1);
//    assert( p[0][i] <= 3);
    cout << p[0][i] << " ";
  } 
  cout << endl;
  for(int i=0; i<N; i++) {
//    assert( p[1][i] >= 1);
//    assert( p[1][i] <= 3);
    cout << p[1][i] << " ";
  } 
  cout << endl;
  return;
}

####

int listaH(int n);
int listaG(int n);

int listaF(int n) {
  // stampa uno ad uno i possibili modi di completare la
  // piastrellatura parziale del bagno 2xN
  // presente nella variabile p. La parte ancora bianca (0) entro p
  // forma un sottobagno ben quadrato 2xn
  assert(n >= 0);
  if(n==0)
    printP();
    return;
  if(n==1)
    int min_free = 1;
    if(min_free == p[0][N-2]) min_free++;
    if(min_free == p[1][N-2]) min_free++;
    if(min_free == p[0][N-2]) min_free++;
    int col2 = min_free+1;
    if(col2 == p[0][N-2]) col2++;
    if(col2 == p[1][N-2]) col2++;
    if(col2 == p[0][N-2]) col2++;
    p[0][N-2] = p[1][N-2] = min_free; 
    printP();
    p[0][N-2] = p[1][N-2] = min_free; 


    printP();    
    return;
  return listaF(n-1) + listaG(n-1) + listaH(n-2);
}

int listaG(int n) {
  assert(n >= 0);
  if(n==0)
    return 1;
  return listaF(n) + listaG(n-1);
}

int listaH(int n) {
  assert(n >= 0);
  return listaF(n) + listaG(n);
}


####

        
int main() {
#ifdef EVAL
        freopen("input.txt", "r", stdin);
        freopen("output.txt", "w", stdout);
#endif
  cin >> N;
  cout << listaF(N) << endl;
  printP();
  return 0;
}

File: classes/RO/2019-2020/path/2020-03-24/piastrelle2.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

def F(n):
    """Ritorna il numero di modi di piastrellare
    un bagno 2xn ben squadrato"""
    assert n >= 0
    if n == 0:
        return 1
    if n == 1:
        return 2
    return F(n-1) + G(n-1) + H(n-2)

def G(n):
    """Ritorna il numero di modi di piastrellare
    un bagno 2x(n+1) che ha perso una cella d'angolo.
    ben squadrato. Nota: la riga corta ha n celle."""
    assert n >= 0
    if n == 0:
        return 1
    return F(n) + G(n-1)

def H(n):
    """Ritorna il numero di modi di piastrellare
    un bagno 2x(n+2) che ha perso due celle di cui precisamente una d'angolo e poi una sua limitrofa. Nota: la riga corta ha n celle."""
    assert n >= 0
    return G(n) + F(n)

############

def stampaPiastrellatureF(n):
    """Stampa tutte le piastrellature del bagno 2xN compatibili con quanto si trova nella variabile globale piastrellatura.
       Resta da scegliere come piastrellare quel sottobagno di celle (i,j)
       con  piastrellatura[i][j] = 0.
       Esso è un bagno 2xn ben squadrato collocato a destra entro il bagno 2xN."""
    assert n >= 0
    if n == 0:
        printPiastrellatura()
        return
    min_free = 1
    assert piastrellatura[0][N-n-1] != 0
    if min_free == piastrellatura[0][N-n-1]:
        min_free += 1
    if min_free == piastrellatura[1][N-n-1]:
        min_free += 1
    if min_free == piastrellatura[0][N-n-1]:
        min_free += 1
    free2 = min_free + 1
    if free2 == piastrellatura[0][N-n-1]:
        free2 += 1
    if free2 == piastrellatura[1][N-n-1]:
        free2 += 1
    if free2 == piastrellatura[0][N-n-1]:
        free2 += 1
    if n == 1:
        piastrellatura[0][N-1] = piastrellatura[1][N-1] = min_free 
        printPiastrellatura()
        piastrellatura[0][N-1] = min_free
        piastrellatura[1][N-1] = free2
        printPiastrellatura()
        return
    piastrellatura[0][N-1] = piastrellatura[1][N-1] = min_free 
    stampaPiastrellatureF(n-1)
    piastrellatura[0][N-1] = piastrellatura[1][N-1] = 0 
    piastrellatura[0][N-1] = min_free 
    stampaPiastrellatureG(n-1)
    piastrellatura[0][N-1] = 0 
    stampaPiastrellatureH(n-2)
    piastrellatura[0][N-1] = piastrellatura[0][N-2] = min_free 

def stampaPiastrellatureG(n):
    """Stampa tutte le piastrellature del bagno 2xN compatibili con quanto si trova nella variabile globale piastrellatura.
       Resta da scegliere come piastrellare quel sottobagno di celle (i,j)
       con  piastrellatura[i][j] = 0.
       Esso è un bagno 2x(n+1) che ha perso una cella d'angolo.
       Nota: la riga corta ha n celle."""
    assert n >= 0
    if n == 0:
        piastrellatura[i][j] = 0
        return 1

    
    return stampaPiastrellatureF(n) + stampaPiastrellatureG(n-1)

def stampaPiastrellatureH(n):
    """Stampa *tutte* le piastrellature di
    un bagno 2x(n+2) che ha perso due celle di cui precisamente una d'angolo e poi una sua limitrofa. Nota: la riga corta ha n celle."""
    assert n >= 0
    return stampaPiastrellatureG(n) + stampaPiastrellatureF(n)

############

with open("input.txt", "r") as fr:
    N = int(fr.readline())

fw = open("output.txt", "w")

def printPiastrellatura():
    for i in range(N):
        fw.write(str(piastrellatura[0][i]) + " ")
    fw.write("\n")
    for i in range(N):
        fw.write(str(piastrellatura[1][i]) + " ")
    fw.write("\n")

fw.write(str(F(N)))
fw.write("\n")

piastrellatura = [ [0]*N, [0]*N]

#printPiastrellatura()
    
stampaPiastrellatureF(N)
